wp-content/plugins/tapestry/controller/class.user-controller.php

```diff
@@ -0,0 +1,140 @@
+     * @param type @postId the post's ID
+     * @param type @nodeId the current node being viewed
+     * @param type @progressValue is how much the video was viewed, value should be between >= 0 and <= 1
+    */
+    public function updateProgress($postId, $nodeId, $progressValue) {
+        $this->_checkUserAndPostId($postId);
+
+        if ($progressValue !== NULL) {
+            $progressValue = floatval($progressValue);
+        }
+
+        // Value should be between 0 and 1
+        if ($progressValue < 0 || $progressValue > 1) {
+            throw new Exception('Invalid progress value');
+        }
+
+        $this->_updateUserProgress($postId, $nodeId, $progressValue);
+    }
+
+    /** 
+     * Get User's video progress for a tapestry post
+     * @param type @postId the post's ID
+     * @param type @nodeIdArr is a list of ids currently in the tapestry
+    */
+    public function getProgress($postId, $nodeIdArr) {
+        $this->_isValidTapestryPost($postId);
+        $this->_checkUserAndPostId($postId);
+
+        return $this->_getUserProgress($postId, $nodeIdArr);
+    }
+
+    /** 
+     * Update User's h5p video setting for a tapestry post
+     * @param type @postId the post's ID
+     * @param type @h5pSettingsData stores volume, playbackRate, quality of h5p video
+    */
+    public function updateH5PSettings($postId, $h5pSettingsData) {
+        $this->_checkUserAndPostId($postId);
+
+        if ($this->_isJson($h5pSettingsData)) {
+            $h5pSettingsData = json_decode($h5pSettingsData);
+        } else {
+            throw new Exception('Invalid json');
+        }
+
+        $this->_updateUserH5PSettings($postId, $h5pSettingsData);
+    }
+
+    /** 
+     * Get User's h5p video setting for a tapestry post
+     * @param type @postId the post's Id
+    */
+    public function getH5PSettings($postId) {
+        $this->_isValidTapestryPost($postId);
+        $this->_checkUserAndPostId($postId);
+        return $this->_getUserH5PSettings($postId);
+    }
+
+    private function _updateUserProgress($postId, $nodeId, $progressValue) {
+        $userId = $this->userId;
+        update_user_meta($userId, 'tapestry_' . $postId . '_progress_node_' . $nodeId, $progressValue);
+    }
+
+    private function _getUserProgress($postId, $nodeIdArr) {
+        $userId = $this->userId;
+        $progress = new stdClass();
+
+        // Build json object for frontend e.g. {0: 0.1, 1: 0.2} where 0 and 1 are the node IDs
+        foreach ($nodeIdArr as $nodeId) {
+            $progress_value = get_user_meta($userId, 'tapestry_' . $postId . '_progress_node_' . $nodeId, true);
+            if ($progress_value !== NULL) {
+                $progress->$nodeId = $progress_value;
+            } else {
+                $progress->$nodeId = 0.0;
+            }
+        }
+
+        return json_encode($progress);
+    }
+
+    private function _updateUserH5PSettings($postId, $h5pSettingsData) {
+        $userId = $this->userId;
```

Comment:
I prefer you not declare new variables if you are using them only once or twice. It'd be easier to read your code if you just plug `$this->userId` directly where you need it. Same goes in other places in your code. Thanks!

Reply:
I prefer it this way for extendability in the future. But will change.